Sources/Features/Shared/CloudKitOperation.swift

```diff
@@ -276,7 +585,97 @@ public class CloudKitRecovery<T where T: NSOperation, T: CKOperationType> {
+ these can be done via their initializers. However, it is also possible to set the properties directly.
+ This can be done directly onto the CloudKitOperation. For example, given the above:
+
+ ```swift
+ let container = CKContainer.defaultContainer()
+ operation.container = container
+ operation.database = container.privateCloudDatabase
+ ```
+
+ ## Completion
+
+ All CKOperation subclasses have a completion block which should be set. This completion block receives
+ the "results" of the operation, and an `NSError` argument. However, CloudKitOperation features its own
+ semi-automatic error handling system. Therefore, the only completion block needed is one which receives
+ the "results". Essentially, all that is needed is to manage the happy path of the operation. For all
+ CKOperation subclasses, this can be configured directly on the CloudKitOperation instance, using a 
+ pattern of `setOperationKindCompletionBlock { }`. For example, given the above:
+
+```swift
+ operation.setFetchRecordZonesCompletionBlock { zonesByID in
+    // Do something with the zonesByID
+ }
+```
+
+Note, that for the automatic error handling to kick in, the happy path must be set (as above).
+
+ ### Error Handling
+
+ When the completion block is set as above, any errors receives from the CKOperation subclass are 
+ intercepted, and instead of the provided block being executed, the operation finsihes with an error.
+
+ However, CloudKitOperation is a subclass of RetryOperation, which is actually a GroupOperation subclass, 
+ and when a child operation finishes with an error, RetryOperation will consult its error handler, and
+ attempt to retry the operation.
+
+ In the case of CloudKitOperation, the error handler has automatic support for many common error kinds.
+ When the CKOperation receives an error, the CKErrorCode is extrated, and the handler consults a 
+ CloudKitRecovery instance to check for a particular way to handle that error code. In some cases, this will
+ result in a tuple being returned. The tuple is an optional delay, and a new instance of the CKOperation class.
+
+ This is why the initializer takes a block which returns an instance of the CKOperation subclass.
+
+ The delay is used to automatically respect any wait periods returned in the CloudKit NSError object.
+
+ If the error recovery does not have a handler, or the handler returns nil (no tuple), the CloudKitOperation 
+ will finish (with the errors).
+
+ ### Custom Error Handling
+
+ To provide bespoke error handling, further configure the CloudKitOperation by calling setErrorHandlerForCode.
+ For example:
+
+ ```swift
+ operation.setErrorHandlerForCode(.PartialFailure) { error, log, suggested in
+    return suggested
+ }
+ ```
+
+ Note that the error handler receives, the received error, a logger object, and the suggested tuple, which
+ could be modified before being returned. Alternatively, return nil to not retry.
+
+*/
```

Comment:
Sounds pretty cool. I'm sure I'd almost always retry. 

Question: in my Objective C version I never fussed a lot with Reachability because I thought that I could use the CloudKit's own error codes to deal with setting up repeated attempts at the operations. Is that accurate?

 

> On Feb 6, 2016, at 18:01, Daniel Thorpe <notifications@github.com> wrote:
> 
> In Sources/Features/Shared/CloudKitOperation.swift:
> 
> > +
> > + ### Custom Error Handling
> > +
> > + To provide bespoke error handling, further configure the CloudKitOperation by calling setErrorHandlerForCode.
> > + For example:
> > +
> > + ```swift
> > + operation.setErrorHandlerForCode(.PartialFailure) { error, log, suggested in
> > +    return suggested
> > + }
> > + ```
> > +
> > + Note that the error handler receives, the received error, a logger object, and the suggested tuple, which
> > + could be modified before being returned. Alternatively, return nil to not retry.
> > +
> > +*/
> @StephenDaedalus does this makes sense?
> 
> —
> Reply to this email directly or view it on GitHub.
>